### PR TITLE
Us041 authorisation system

### PR DESCRIPTION
**Pull request by Fluffyguy611**
### 🛠 Changes being made

Added authorisation System

## install following npm dependenties
```
npm install cookie-parser --save
```
```
npm install express-session --save
```
```
 npm i --save-dev @types/cookie-parser   
```
```
npm install express-cookie --save
```

### CODE REVIEW

File: product-backend/src/main/java/org/kainos/ea/controller/AuthController.java [hunk 1]
```diff
@@ -12,12 +12,18 @@
...
 import org.kainos.ea.service.AuthService;
 import org.kainos.ea.service.AuthValidator;
 import org.kainos.ea.service.PasswordService;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
```
Comment (ljercha) [inline product-backend/src/main/java/org/kainos/ea/controller/AuthController.java]: cleanup imports

File: product-backend/src/main/java/org/kainos/ea/controller/RestrictedOperationsRequestFilter.java [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+import java.sql.SQLException;
+import java.util.Optional;
+import java.util.Properties;
+
+@Provider
+@PreMatching
+@Priority(Priorities.AUTHORIZATION)
+public class RestrictedOperationsRequestFilter implements ContainerRequestFilter {
+
+    private final static Logger logger = LoggerFactory.getLogger(CapabilityService.class);
```
Comment (ljercha) [inline product-backend/src/main/java/org/kainos/ea/controller/RestrictedOperationsRequestFilter.java]:
```suggestion
    private final static Logger logger = LoggerFactory.getLogger(RestrictedOperationsRequestFilter.class);
```

File: product-backend/src/main/java/org/kainos/ea/controller/RestrictedOperationsRequestFilter.java [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+
+
+        }catch (FailedToVerifyTokenException e){
+            ctx.abortWith(Response.status(Response.Status.NOT_FOUND).entity(e.getMessage()).build());
+            logger.info("Failed to Verify Token");
+        }
+    }
+
+    public DecodedJWT getDecodedJWT(ContainerRequestContext ctx) throws FailedToVerifyTokenException {
+        Algorithm algorithm = Algorithm.HMAC256("erggv45wv54c53xd345vcg4v54yv2");
```
Comment (ljercha) [inline product-backend/src/main/java/org/kainos/ea/controller/RestrictedOperationsRequestFilter.java]: Extract "erggv45wv54c53xd345vcg4v54yv2" to shared static variable.
```suggestion
        Algorithm algorithm = Algorithm.HMAC256(Secrets.TOKEN_SECRET);
```

File: product-backend/src/test/java/AuthServiceTest.java [hunk 4]
```diff
@@ -97,9 +99,9 @@ void loginUser_When_UserIsValid_Expect_TokenReturned() throws SQLException, Inva
         when(passwordServiceMock.verifyHash(mockedLogin.getPassword(), existingUserMock.get().getPassword())).thenReturn(true);
         when(authDaoMock.getUserByEmail(mockedLogin.getEmail(), databaseConnectorMock.getConnection())).thenReturn(existingUserMock);
 
-        String token = authService.login(mockedLogin);
+        UserCredentials token = authService.login(mockedLogin);
 
-        assertThat(token).isNotEmpty();
+        assertThat(token).isNotNull();
```
Comment (ljercha) [inline product-backend/src/test/java/AuthServiceTest.java]: Update assertions
- validate token is long string
- roleId is passed to the returned value

File: product-web-app/middleware/auth.ts [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+import axios from 'axios';
+
+export default class AuthMiddleware {
+  appRoutes(app: Application) {
+    app.use((req: Request, res: Response, next: NextFunction) => {
+      if (req.path === '/auth/register') {
+        next();
+      } else if (req.cookies.token) {
+        axios.defaults.headers.common.Authorization = req.cookies.token;
+        req.session.isAdmin = req.cookies.roleId;
```
Comment (ljercha) [inline product-web-app/middleware/auth.ts]: Can you try to use
app.locals
http://expressjs.com/en/api.html
They should be automatically added to the render function and available in the nunjuck templates

**Review by ljercha — COMMENTED** [review]
Try to use app.locals to simplify passing flag to templates
http://expressjs.com/en/api.html#app.locals

File: product-backend/src/main/java/org/kainos/ea/exception/FailedToVerifyTokenException.java [hunk 6]
```diff
@@ -0,0 +1,8 @@
+package org.kainos.ea.exception;
+
+public class FailedToVerifyTokenException extends Throwable {
+    @Override
+    public String getMessage() {
+        return "Failed to Verify Token";
```
Comment (mtslech) [inline product-backend/src/main/java/org/kainos/ea/exception/FailedToVerifyTokenException.java]:
```suggestion
        return "Failed to verify token";
```

File: product-backend/src/main/java/org/kainos/ea/exception/NoActiveTokenException.java [hunk 7]
```diff
@@ -0,0 +1,5 @@
+package org.kainos.ea.exception;
+
+public class NoActiveTokenException extends Throwable {
+
+}
```
Comment (mtslech) [inline product-backend/src/main/java/org/kainos/ea/exception/NoActiveTokenException.java]:
```suggestion
package org.kainos.ea.exception;

public class NoActiveTokenException extends Throwable {
@Override
    public String getMessage() {
        return "No active token";
    }
}
```

File: product-web-app/views/parent.html
Comment (mtslech) [inline product-web-app/views/parent.html:1]: I'm also working on it on branch feature/us026/add-branding simultaneously with same features, just FYI

**Review by ljercha — APPROVED** [review]

**Review by mtslech — APPROVED** [review]

**Pull request merged**